sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java

```diff
@@ -109,30 +113,121 @@ public abstract class RowCoderGenerator {
+      this.stackTrace = stackTrace;
+    }
+
+    public T getValue() {
+      return value;
+    }
+
+    public String getStackTrace() {
+      return stackTrace;
+    }
+  }
+
+  // Cache for Coder class that are already generated. Coders are added with setOverridesLock held.
+  private static final Map<UUID, WithStackTrace<Coder<Row>>> GENERATED_CODERS =
       Maps.newConcurrentMap();
 
+  @GuardedBy("setOverridesLock")
+  private static final Map<UUID, WithStackTrace<Map<String, Integer>>> ENCODING_POSITION_OVERRIDES =
+      Maps.newHashMap();
+
+  private static final Object setOverridesLock = new Object();
+
+  private static final Logger LOG = LoggerFactory.getLogger(RowCoderGenerator.class);
+
+  private static String getStackTrace() {
+    StringBuilder builder = new StringBuilder();
+    for (StackTraceElement e : new Throwable().getStackTrace()) {
+      builder.append("  at ").append(e).append("\n");
+    }
+    return builder.toString();
+  }
+
   public static void overrideEncodingPositions(UUID uuid, Map<String, Integer> encodingPositions) {
-    ENCODING_POSITION_OVERRIDES.put(uuid, encodingPositions);
+    final String stackTrace = getStackTrace();
+    synchronized (setOverridesLock) {
+      @Nullable
+      WithStackTrace<Map<String, Integer>> previousEncodingPositions =
+          ENCODING_POSITION_OVERRIDES.put(
+              uuid, new WithStackTrace<>(encodingPositions, stackTrace));
+      @Nullable WithStackTrace<Coder<Row>> existingCoder = GENERATED_CODERS.get(uuid);
+      if (previousEncodingPositions == null) {
+        if (existingCoder != null) {
+          LOG.error(
+              "Received encoding positions for uuid {} too late after creating RowCoder. Created: {}\n Override: {}",
+              uuid,
+              existingCoder.getStackTrace(),
+              stackTrace);
+        } else {
+          LOG.info("Received encoding positions {} for uuid {}.", encodingPositions, uuid);
+        }
+      } else if (!previousEncodingPositions.getValue().equals(encodingPositions)) {
+        if (existingCoder == null) {
+          LOG.error(
+              "Received differing encoding positions for uuid {} before coder creation. Was {} at {}\n Now {} at {}",
+              uuid,
+              previousEncodingPositions.getValue(),
+              encodingPositions,
+              previousEncodingPositions.getStackTrace(),
+              stackTrace);
+        } else {
+          LOG.error(
+              "Received differing encoding positions for uuid {} after coder creation at {}\n. "
+                  + "Was {} at {}\n Now {} at {}\n",
+              uuid,
+              existingCoder.getStackTrace(),
+              previousEncodingPositions.getValue(),
+              encodingPositions,
+              previousEncodingPositions.getStackTrace(),
+              stackTrace);
+        }
+      }
+    }
+  }
+
+  @VisibleForTesting
+  static void clearRowCoderCache() {
+    synchronized (setOverridesLock) {
```

Comment:
Coders can be read without the synchronized block but are written with the extra synchronization because concurrenthashmap updateifnotpresent is not reentrant (unlike synchronized) and the get/insert pattern is possibly racy.

That said I since the read coders are cached, I think I will just change to regular maps under synchronization and ditch the concurrenthashmap.